File: git-epic-creator/services/code_graph_ingestion_service/src/core/workspace.py
"""Deterministic workspace layout helpers.

The ingestion pipeline materializes repos into a project-scoped workspace on disk.
Paths must be deterministic for identical inputs (CGI-FR-005).
"""

from __future__ import annotations

import hashlib
import re
import shutil
from dataclasses import dataclass
from pathlib import Path


_SAFE_SEGMENT_RE = re.compile(r"^[A-Za-z0-9][A-Za-z0-9_.-]{0,127}$")


def _stable_key(value: str) -> str:
    """Return a stable directory-safe key.

    Prefer a readable safe segment; otherwise fall back to a short sha256 prefix.
    """

    if _SAFE_SEGMENT_RE.match(value):
        return value
    h = hashlib.sha256(value.encode("utf-8", errors="replace")).hexdigest()
    return f"sha256-{h[:16]}"


@dataclass(frozen=True)
class Workspace:
    """Workspace layout rooted at a deterministic base directory."""

    root: Path

    def project_dir(self, project_id: str) -> Path:
        return self.root / _stable_key(project_id)

    def zip_repo_dir(self, project_id: str) -> Path:
        return self.project_dir(project_id) / "zip" / "repo"

    def git_repo_dir(self, project_id: str, git_url: str, ref: str | None) -> Path:
        key_src = f"{git_url}\n{ref or ''}"
        key = _stable_key(key_src)
        return self.project_dir(project_id) / "git" / key / "repo"

    def ensure_empty_dir(self, path: Path) -> None:
        """Ensure `path` exists and is empty (deterministic materialization)."""

        if path.exists():
            shutil.rmtree(path)
        path.mkdir(parents=True, exist_ok=True)

File: git-epic-creator/services/code_graph_ingestion_service/src/core/workspace.rs
use sha2::{Digest, Sha256};
use std::path::{Path, PathBuf};

fn is_safe_segment(s: &str) -> bool {
    let bytes = s.as_bytes();
    if bytes.is_empty() || bytes.len() > 128 {
        return false;
    }
    let first = bytes[0];
    if !first.is_ascii_alphanumeric() {
        return false;
    }
    bytes.iter().all(|b| {
        b.is_ascii_alphanumeric() || matches!(*b, b'_' | b'.' | b'-')
    })
}

fn stable_key(value: &str) -> String {
    if is_safe_segment(value) {
        return value.to_string();
    }
    let mut h = Sha256::new();
    h.update(value.as_bytes());
    let digest = hex::encode(h.finalize());
    format!("sha256-{}", &digest[..16])
}

#[derive(Debug, Clone)]
pub struct Workspace {
    pub root: PathBuf,
}

impl Workspace {
    pub fn new(root: impl AsRef<Path>) -> Self {
        Self {
            root: root.as_ref().to_path_buf(),
        }
    }

    pub fn project_dir(&self, project_id: &str) -> PathBuf {
        self.root.join(stable_key(project_id))
    }

    pub fn zip_repo_dir(&self, project_id: &str) -> PathBuf {
        self.project_dir(project_id).join("zip").join("repo")
    }

    pub fn git_repo_dir(&self, project_id: &str, git_url: &str, ref_name: Option<&str>) -> PathBuf {
        let key_src = format!("{}\n{}", git_url, ref_name.unwrap_or(""));
        let key = stable_key(&key_src);
        self.project_dir(project_id).join("git").join(key).join("repo")
    }
}


